lighthouse-core/lib/page-functions.js

```diff
@@ -446,20 +446,60 @@ function wrapRequestIdleCallback(cpuSlowdownMultiplier) {
   };
 }
 
-const getNodeDetailsString = `function getNodeDetails(elem) {
+/**
+ * @param {HTMLElement} element
+ */
+function getNodeDetailsImpl(element) {
+  element = element instanceof ShadowRoot ? element.host : element;
+
+  // This bookkeeping is for the FullPageScreenshot gatherer.
+  if (!window.__lighthouseNodesDontTouchOrAllVarianceGoesAway) {
+    window.__lighthouseNodesDontTouchOrAllVarianceGoesAway = {
+      lhIdToElements: new Map(),
+      elementToLhId: new Map(),
+    };
+  }
+
+  // Create an id that will be unique across all execution contexts.
+  // The id could be any arbitrary string, the exact value is not important.
+  // For example, tagName is added only because it might be useful for debugging.
+  // But execution id and map size are added to ensure uniqueness.
+  // We also dedupe this id so that details collected for an element within the same
+  // pass and execution context will share the same id. Not technically important, but
+  // cuts down on some duplication.
+  let lhId = window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.elementToLhId.get(element);
+  if (!lhId) {
+    lhId = [
+      window.__lighthouseExecutionContextId !== undefined ?
+        window.__lighthouseExecutionContextId :
+        'page',
+      window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.elementToLhId.size,
+      element.tagName,
+    ].join('-');
+    window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.lhIdToElements.set(lhId, element);
```

Comment:
seems like we don't need this.  the `for (const [id, node] of lhIdToElements.entries()) {` later on could just use `elementToLhId` and reverse the order of the destructured array. right?

Reply:
ya but I'm assuming not performant. and this code is simpler/fewer lines.

Reply:
> ya but I'm assuming not performant. and this code is simpler/fewer lines.

hmmm.. maybe you're thinking of something else?
what i'm proposing is definitely less work.  and its fewer lines


```patch
diff --git a/lighthouse-core/gather/gatherers/full-page-screenshot.js b/lighthouse-core/gather/gatherers/full-page-screenshot.js
index 474eca578..5806de603 100644
--- a/lighthouse-core/gather/gatherers/full-page-screenshot.js
+++ b/lighthouse-core/gather/gatherers/full-page-screenshot.js
@@ -82,9 +82,8 @@ class FullPageScreenshot extends Gatherer {
       /** @type {LH.Artifacts.FullPageScreenshot['nodes']} */
       const nodes = {};
       if (!window.__lighthouseNodesDontTouchOrAllVarianceGoesAway) return nodes;
-
-      const {lhIdToElements} = window.__lighthouseNodesDontTouchOrAllVarianceGoesAway;
-      for (const [id, node] of lhIdToElements.entries()) {
+      const elementToLhId = window.__lighthouseNodesDontTouchOrAllVarianceGoesAway;
+      for (const [node, id] of elementToLhId.entries()) {
         // @ts-expect-error - getBoundingClientRect put into scope via stringification
         const rect = getBoundingClientRect(node);
         if (rect.width || rect.height) nodes[id] = rect;
diff --git a/lighthouse-core/lib/page-functions.js b/lighthouse-core/lib/page-functions.js
index 9769877a9..a71bfe497 100644
--- a/lighthouse-core/lib/page-functions.js
+++ b/lighthouse-core/lib/page-functions.js
@@ -452,10 +452,7 @@ function wrapRequestIdleCallback(cpuSlowdownMultiplier) {
 function getNodeDetailsImpl(element) {
   // This bookkeeping is for the FullPageScreenshot gatherer.
   if (!window.__lighthouseNodesDontTouchOrAllVarianceGoesAway) {
-    window.__lighthouseNodesDontTouchOrAllVarianceGoesAway = {
-      lhIdToElements: new Map(),
-      elementToLhId: new Map(),
-    };
+    window.__lighthouseNodesDontTouchOrAllVarianceGoesAway = new Map();
   }
 
   // Create an id that will be unique across all execution contexts.
@@ -465,17 +462,16 @@ function getNodeDetailsImpl(element) {
   // We also dedupe this id so that details collected for an element within the same
   // pass and execution context will share the same id. Not technically important, but
   // cuts down on some duplication.
-  let lhId = window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.elementToLhId.get(element);
+  let lhId = window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.get(element);
   if (!lhId) {
     lhId = [
       window.__lighthouseExecutionContextId !== undefined ?
         window.__lighthouseExecutionContextId :
         'page',
-      window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.elementToLhId.size,
+      window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.size,
       element.tagName,
     ].join('-');
-    window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.lhIdToElements.set(lhId, element);
-    window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.elementToLhId.set(element, lhId);
+    window.__lighthouseNodesDontTouchOrAllVarianceGoesAway.set(element, lhId);
   }
 
   const htmlElement = element instanceof ShadowRoot ? element.host : element;
diff --git a/types/externs.d.ts b/types/externs.d.ts
index a04a09b08..a4727bb45 100644
--- a/types/externs.d.ts
+++ b/types/externs.d.ts
@@ -324,11 +324,8 @@ declare global {
   }
 
   interface Window {
-    /** Used by FullPageScreenshot gatherer. */
-    __lighthouseNodesDontTouchOrAllVarianceGoesAway: {
-      lhIdToElements: Map<string, HTMLElement>;
-      elementToLhId: Map<HTMLElement, string>;
-    };
+    /** Map of Element to lhID. Used by FullPageScreenshot gatherer. */
+    __lighthouseNodesDontTouchOrAllVarianceGoesAway: Map<HTMLElement, string>;
     __lighthouseExecutionContextId?: number;
   }
 }

```

Reply:
ya this is good. i did not think to do just a `Map<HTMLElement, string>`, was stuck on thinking of it as `Map<string, HTMLElement>`